FINERACT-2259: Manage external IDs-Savings

Lymah123 opened the pull request:
# Description

## Introduction

Transactions may change due to correction. Correction can be done by reverting a transaction. This can lead to removing and adding transactions back to an account. It results new ID distribution from the transaction sequence. External systems may not want to follow up on these changes but rather access transactions as they are at their end state. In order to be able to direct client immediately to the right record an additional identifier is to be used.

Supporting externalId is a long time wanted functionality with two major benefits:

- Not an auto incremented number (privacy, harder to guess)
- Not represent any internal state

## Requirements

🟩 Supporting external id for identifying and fetching some if the entities in Fineract
🟩 API enhancement to support targeting by external id
✅ External id can be provided in API request
🟩 Extend API response with external id
🟩 Support external id auto generation (configurable)

**External identifier**

- Must be unique
- Immutable
- Customer (Caller) provided
- Ability to generate automatically (configurable)

**Database**
External identifier field to be stored in the loan table

_type_: String

_indexing_: true

## Configuration

Enable to auto generate external id (if not provided).

By default it should be disabled.

_**This configuration already exists!**_

 

## Goal

This story is focusing to introduce external ids for savings accounts and transactions!

Lymah123 committed:
FINERACT-2259: Manage external IDs-Savings

adamsaghy commented:
@Lymah123 Please run this before any commits:
`./gradlew spotlessApply spotbugsMain spotbugsTest checkstyleMain checkstyleTest
`

Lymah123 commented:
> @Lymah123 Please run this before any commits:
> `./gradlew spotlessApply spotbugsMain spotbugsTest checkstyleMain checkstyleTest
> `

Ok.

Lymah123 commented:
Hi @adamsaghy 

I've tried building it on my end, but some call sites require externalId implementations, which I'm currently working on. I'll push the updates once I've completed and tested the changes.

I just thought of updating you on the progress.

Thank you.

adamsaghy commented:
```
Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.
> A failure occurred while executing org.nosphere.apache.rat.RatWork
   > Apache Rat audit failure - 4 unapproved licenses

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.
     	See file:///home/runner/work/fineract/fineract/build/reports/rat/index.html
 ```
 
 Please make sure all new files has the apache license header! (you can copy from any existing file)

Lymah123 commented:
> ```
> Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.
> > A failure occurred while executing org.nosphere.apache.rat.RatWork
>    > Apache Rat audit failure - 4 unapproved licenses
> 
> You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.
>      	See file:///home/runner/work/fineract/fineract/build/reports/rat/index.html
> ```
> 
> Please make sure all new files has the apache license header! (you can copy from any existing file)

@adamsaghy , I have worked on the files.

adamsaghy commented:
you can see all the checks are red!

`home/runner/work/fineract/fineract/fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/DepositAccountWritePlatformServiceJpaRepositoryImpl.java:366: error: method handleFDDeposit in interface DepositAccountDomainService cannot be applied to given types;
2025-05-12T15:46:58.1104587Z         final SavingsAccountTransaction deposit = this.depositAccountDomainService.handleFDDeposit(account, fmt, transactionDate,
2025-05-12T15:46:58.1106178Z                                                                                   ^
2025-05-12T15:46:58.1106991Z > Task :fineract-provider:compileJava
2025-05-12T15:46:58.1108324Z   required: FixedDepositAccount,DateTimeFormatter,LocalDate,BigDecimal,PaymentDetail,ExternalId
2025-05-12T15:46:58.1109870Z   found:    FixedDepositAccount,DateTimeFormatter,LocalDate,BigDecimal,PaymentDetail
2025-05-12T15:46:58.1111070Z   reason: actual and formal argument lists differ in length
2025-05-12T15:46:58.1113719Z `

Lymah123 commented:
Hi @adamsaghy, I have made some changes and tried building and testing locally, which was successful.

adamsaghy commented:
@Lymah123 Please rebase your PR with the latest develop branch!

Also make sure you have pushed all the changes. The above listed errors are compilation errors...

a7med3del1973 commented:
Please, solve the conflict.

Lymah123 commented:
I'm experiencing an unusual conflict situation with my PR.

GitHub shows conflicts in three files:
```
fineract-provider/build.gradle
fineract-provider/src/main/resources/jpa/persistence.xml
fineract-savings/src/main/resources/jpa/savings/persistence.xml
```
However, my local Git shows no conflicts:

- When I run `git merge develop` → "Already up to date"
- When I run `git rebase develop` → "Current branch is up to date"
- I created a fresh branch from develop and cherry-picked my changes without conflicts

What I've Tried
1. Force pushing with `--force-with-lease` and `--force`
2. Creating new branches from develop
3. Hard resetting and cherry-picking
4. Creating a clean branch with `conflict-resolution-branch`
   - This cherry-picked cleanly with all 22 files
   - But GitHub still shows "Can't automatically merge"

What approach would you recommend to resolve these "phantom" conflicts?

adamsaghy commented:
There is no such things as phantom conflicts.. The recommended way is to rebase your feature branch. However what you have tried to do was to rebase the local develop branch, however if you didnt pulled the latest changes of develop branch from the remote, it will not do much things, so let me advise 2 options what you can do:

Option 1:
- `git checkout develop` -> Switch to `develop` branch
- `git pull` -> Pull latest changes from remote to local `develop` branch
- `git checkout external-id-management` -> Switch to your feature branch
- `git rebase develop` -> Start rebasing -> Fix conflicts
- `git rebase --continue` -> Finish on Rebasing

Option 2:
- `git fetch origin` -> Fetch latest branches from `origin` (if `origin` points to `apache/fineract` repository)
- `git rebase origin/develop`  -> Start rebasing -> Fix conflicts
- `git rebase --continue` -> Finish on Rebasing

Lymah123 commented:
The issue wasn't actually "phantom" conflicts , thanks for the correction, @adamsaghy 
I think it was a remote repository mismatch

- **Local setup**: My ` origin` remote pointed to my fork `(Lymah123/fineract)`
- **GitHub PR**: The PR was targeting the upstream repository `(apache/fineract)`
- **Problem**: My local `develop` branch was synced with my fork, but GitHub was comparing against the upstream repository's `develop` branch

Why Local Git Showed No Conflicts when I run:

```
git merge develop     
git rebase develop
```

These commands were comparing against my fork's `develop` branch, which was already compatible with my `feature` branch.

The Real Issue: 
GitHub's conflict detection was comparing my PR branch against upstream's `develop` branch, which had diverged from my fork's `develop` branch.

Steps I took to resolved it:
1. Added upstream remote
```
git remote add upstream https://github.com/apache/fineract.git
git fetch upstream
```
2. Rebased against the correct target
```
git checkout external-id-management
git rebase upstream/develop  # This revealed the real conflicts

```

3. Resolved conflicts and updated PR
```
git add <conflicted-files>
git rebase --continue
git push origin external-id-management --force-with-lease
```

Key Takeaway
I will always rebase feature branches against the upstream repository's target branch `(upstream/develop)`, not my fork's copy, especially on `apache/fineract` repository.

My PR is ready for review @adamsaghy. Thanks!

adamsaghy commented:
```
/home/runner/work/fineract/fineract/fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/DepositAccountWritePlatformServiceJpaRepositoryImpl.java:366: error: method handleFDDeposit in interface DepositAccountDomainService cannot be applied to given types;
        final SavingsAccountTransaction deposit = this.depositAccountDomainService.handleFDDeposit(account, fmt, transactionDate,
                                                                                  ^
  required: FixedDepositAccount,DateTimeFormatter,LocalDate,BigDecimal,PaymentDetail,ExternalId
  found:    FixedDepositAccount,DateTimeFormatter,LocalDate,BigDecimal,PaymentDetail
  reason: actual and formal argument lists differ in length

adamsaghy commented on config/docker/env/aws.env:
```diff
@@ -17,10 +17,10 @@
 AWS_EC2_METADATA_DISABLED=true
-AWS_ENDPOINT_URL=http://localstack:4666
-#FINERACT_AWS_ENDPOINT=http://localstack:4666
+# AWS_ENDPOINT_URL is deprecated, use FINERACT_AWS_ENDPOINT instead
```
out of scope changes...

Lymah123 replied:
This is for the previous PR I opened. I will delete the branch.

adamsaghy replied:
Still not reverted...

adamsaghy commented on fineract-provider/build.gradle:
```diff
@@ -417,3 +417,5 @@ tasks.register('devRun', org.springframework.boot.gradle.tasks.run.BootRun) {
     }
 }
+
+compileJava.dependsOn(':fineract-core:compileJava')
```
No need...

Lymah123 replied:
Alright, I changed it so I can build locally on my end. I will remove it

adamsaghy commented on fineract-provider/src/main/java/org/apache/fineract/interoperation/service/InteropServiceImpl.java:
```diff
@@ -392,7 +393,7 @@ public InteropTransferResponseData prepareTransfer(@NotNull JsonCommand command)
                     transferCode, null);
             SavingsAccountTransaction holdTransaction = SavingsAccountTransaction.holdAmount(savingsAccount, savingsAccount.office(),
-                    paymentDetail, transactionDate, Money.of(savingsAccount.getCurrency(), total), false);
+                    paymentDetail, transactionDate, Money.of(savingsAccount.getCurrency(), total), false, ExternalId.empty());
```
We should check whether `enable-auto-generated-external-id` is true or false... 

if it is true, we need to generate automatically external id.

Lymah123 replied:
Ok, I will work on that.

adamsaghy replied:
No changes were made...

adamsaghy commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/account/service/AccountTransfersWritePlatformServiceImpl.java:
```diff
@@ -126,13 +126,15 @@ public CommandProcessingResult create(final JsonCommand command) {
                     isRegularTransaction, fromSavingsAccount.isWithdrawalFeeApplicableForTransfer(), isInterestTransfer, isWithdrawBalance);
             final SavingsAccountTransaction withdrawal = this.savingsAccountDomainService.handleWithdrawal(fromSavingsAccount, fmt,
-                    transactionDate, transactionAmount, paymentDetail, transactionBooleanValues, backdatedTxnsAllowedTill);
+                    transactionDate, transactionAmount, paymentDetail, transactionBooleanValues, backdatedTxnsAllowedTill,
```
We should check whether enable-auto-generated-external-id is true or false...

if it is true, we need to generate automatically external id.

Example: 
```
            if (externalId.isEmpty() && TemporaryConfigurationServiceContainer.isExternalIdAutoGenerationEnabled()) {
                externalId = ExternalId.generate();
            }
```
org.apache.fineract.portfolio.loanaccount.service.LoanAssemblerImpl#updateFrom, line 537

adamsaghy replied:
No changes...

adamsaghy commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/FixedDepositAccount.java:
```diff
@@ -197,7 +197,7 @@ public void updateMaturityDateAndAmountBeforeAccountActivation(final MathContext
         final Money transactionAmountMoney = Money.of(getCurrency(), this.accountTermAndPreClosure.depositAmount());
         final SavingsAccountTransaction transaction = SavingsAccountTransaction.deposit(null, office(), null,
-                this.accountSubmittedOrActivationDate(), transactionAmountMoney, refNo);
+                this.accountSubmittedOrActivationDate(), transactionAmountMoney, refNo, ExternalId.empty());
```
We should check whether enable-auto-generated-external-id is true or false...

if it is true, we need to generate automatically external id.

adamsaghy commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/RecurringDepositAccount.java:
```diff
@@ -394,7 +394,7 @@ private List<SavingsAccountTransaction> getTransactions(final LocalDate depositE
                     }
                     final SavingsAccountTransaction transaction = SavingsAccountTransaction.deposit(null, office(), null, dueDate,
-                            installment.getDepositAmountOutstanding(getCurrency()), refNo);
+                            installment.getDepositAmountOutstanding(getCurrency()), refNo, ExternalId.empty());
```
We should check whether enable-auto-generated-external-id is true or false...

if it is true, we need to generate automatically external id.

adamsaghy commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/DepositAccountWritePlatformServiceJpaRepositoryImpl.java:
```diff
@@ -171,7 +171,7 @@ public CommandProcessingResult activateFDAccount(final Long savingsId, final Jso
                     final PaymentDetail paymentDetail = null;
                     this.depositAccountDomainService.handleFDDeposit(account, fmt, account.getActivationDate(),
-                            amountForDeposit.getAmount(), paymentDetail);
+                            amountForDeposit.getAmount(), paymentDetail, ExternalId.empty());
```
We should check whether enable-auto-generated-external-id is true or false...

if it is true, we need to generate automatically external id.

adamsaghy commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsAccountWritePlatformServiceJpaRepositoryImpl.java:
```diff
@@ -244,7 +245,7 @@ public void processPostActiveActions(final SavingsAccount account, final DateTim
             boolean isAccountTransfer = false;
             this.savingsAccountDomainService.handleDeposit(account, fmt, account.getActivationDate(), amountForDeposit.getAmount(), null,
-                    isAccountTransfer, isRegularTransaction, false);
+                    isAccountTransfer, isRegularTransaction, false, ExternalId.empty());
```
We should check whether enable-auto-generated-external-id is true or false...

if it is true, we need to generate automatically external id.

adamsaghy commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/shareaccounts/service/ShareAccountSchedularServiceImpl.java:
```diff
@@ -44,7 +45,7 @@ public void postDividend(final Long dividendDetailId, final Long savingsId) {
         final SavingsAccount savingsAccount = this.savingsAccountAssembler.assembleFrom(savingsId, false);
         SavingsAccountTransaction savingsAccountTransaction = this.savingsAccountDomainService.handleDividendPayout(savingsAccount,
-                DateUtils.getBusinessLocalDate(), shareAccountDividendDetails.getAmount(), false);
+                DateUtils.getBusinessLocalDate(), shareAccountDividendDetails.getAmount(), false, ExternalId.empty());
```
We should check whether enable-auto-generated-external-id is true or false...

if it is true, we need to generate automatically external id.

adamsaghy commented on fineract-savings/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccount.java:
```diff
@@ -549,10 +549,10 @@ public void postInterest(final MathContext mc, final LocalDate interestPostingUp
 
                         newPostingTransaction = SavingsAccountTransaction.interestPosting(this, office(), interestPostingTransactionDate,
-                                interestEarnedToBePostedForPeriod, interestPostingPeriod.isUserPosting());
+                                interestEarnedToBePostedForPeriod, interestPostingPeriod.isUserPosting(), ExternalId.empty());
```
We should check whether enable-auto-generated-external-id is true or false...

if it is true, we need to generate automatically external id.

adamsaghy commented on gradlew:
```diff
@@ -168,36 +168,36 @@ fi
-    APP_HOME=$( cygpath --path --mixed "$APP_HOME" )
-    CLASSPATH=$( cygpath --path --mixed "$CLASSPATH" )
-
+# Convert JAVACMD differently for Cygwin and MSYS
```
Unnecessary changes...

Lymah123 replied:
Ok, this is for my local build. I will remove it before pushing next time.

adamsaghy commented on gradlew.original:
```diff
@@ -0,0 +1,249 @@
+#!/bin/sh
+
+#
```
what is this file?

Lymah123 replied:
For local build, I will remove it.

adamsaghy reviewed (changes requested):
Please see my review and failing tests!

Lymah123 commented:
> Please see my review and failing tests!

I have, and I will work on the requested changes.

Lymah123 commented:
I am still working on it. I wasn't available for some times.

adamsaghy commented on fineract-provider/src/main/java/org/apache/fineract/interoperation/service/InteropServiceImpl.java:
```diff
@@ -417,6 +420,15 @@ public InteropTransferResponseData commitTransfer(@NotNull JsonCommand command)
         String transferCode = request.getTransferCode();
 
+        // Check if auto-generation of external ID is enabled
+        final boolean isAutoGeneratedExternalIdEnabled = this.configurationDomainService.isExternalIdAutoGenerationEnabled();
```
an IF-ELSE would be nicer here...

adamsaghy commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/account/service/AccountTransfersWritePlatformServiceImpl.java:
```diff
@@ -270,6 +274,17 @@ private void undoTransactions(final List<AccountTransferTransaction> accountTran
+     * Returns an appropriate ExternalId based on configuration settings. If auto-generation is enabled and the provided
+     * externalId is empty, generates a new one.
+     */
+    private ExternalId getExternalId(ExternalId externalId) {
```
We dont need this... a simple IF-ELSE would be enought to either generate a new or return empty...

No need for this, use `ExternalIdFactory`

adamsaghy commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/FixedDepositAccount.java:
```diff
@@ -83,6 +83,21 @@ public class FixedDepositAccount extends SavingsAccount {
+     * Returns an appropriate ExternalId based on configuration settings. If auto-generation is enabled and the provided
+     * externalId is empty, generate a new one.
+     */
+    public ExternalId getExternalId(ExternalId externalId) {
```
No need for this, use `ExternalIdFactory`

adamsaghy commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/RecurringDepositAccount.java:
```diff
@@ -91,6 +93,24 @@ public class RecurringDepositAccount extends SavingsAccount {
+     * Returns an appropriate ExternalId based on configuration settings. If auto-generation is enabled and provided
+     * externalId is empty, generate a new one.
+     */
+    public ExternalId getExternalId(ExternalId externalId) {
```
No need for this, use `ExternalIdFactory`

adamsaghy commented on fineract-savings/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccount.java:
```diff
@@ -344,6 +344,17 @@ public class SavingsAccount extends AbstractAuditableWithUTCDateTimeCustom<Long>
 
     public transient ConfigurationDomainService configurationDomainService;
 
+    public ExternalId getExternalId(ExternalId externalId) {
```
No need for this, use `ExternalIdFactory`

adamsaghy commented on fineract-savings/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccount.java:
```diff
@@ -3187,7 +3204,7 @@ private SavingsAccountTransaction handlePayChargeTransactions(SavingsAccountChar
             boolean backdatedTxnsAllowedTill) {
         final SavingsAccountTransaction chargeTransaction = SavingsAccountTransaction.waiver(this, office(),
-                DateUtils.getBusinessLocalDate(), transactionAmount);
+                DateUtils.getBusinessLocalDate(), transactionAmount, null, ExternalId.empty());
```
Rewrite to get the external id as parameter

adamsaghy commented:
@Lymah123 You need to rewrite the whole PR:
- Use `ExternalIdFactory` in the services to generate an external id. It will check whether new should be generated or not.
- Entities and DTOs should get the resolved ExternalId as parameter, no need any further logic in them!
- Hardcoding `ExternalId.empty()` is usually considered incorrect! Use `ExternalIdFactory` instead to get the appropriate ExternalId object and value!

Lymah123 commented:
> @Lymah123 You need to rewrite the whole PR:
> 
> * Use `ExternalIdFactory` in the services to generate an external id. It will check whether new should be generated or not.
> * Entities and DTOs should get the resolved ExternalId as parameter, no need any further logic in them!
> * Hardcoding `ExternalId.empty()` is usually considered incorrect! Use `ExternalIdFactory` instead to get the appropriate ExternalId object and value!

alright @adamsaghy

Lymah123 commented:
Hi @adamsaghy, I have gone through all the feedback. I will work on the requested changes. Thanks.

adamsaghy commented:
Additional issues:
- Liquibase scripts are missing (to add the new column to the savings account transaction database table)
- Testing is missing
- Adding the new field to the supported parameter list is missing
- Validation whether the provided external id length is valid is missing

Lymah123 commented:
> Additional issues:
> 
> * Liquibase scripts are missing (to add the new column to the savings account transaction database table)
> * Testing is missing
> * Adding the new field to the supported parameter list is missing
> * Validation whether the provided external id length is valid is missing

@adamsaghy are these additional issues to work on regards this PR?

IOhacker commented:
@Lymah123 are you still working on this PR?

Lymah123 commented:
> @Lymah123 are you still working on this PR?

Yes @IOhacker,

Lymah123 commented:
> > Additional issues:
> > 
> > * Liquibase scripts are missing (to add the new column to the savings account transaction database table)
> > * Testing is missing
> > * Adding the new field to the supported parameter list is missing
> > * Validation whether the provided external id length is valid is missing
> 
> @adamsaghy are these additional issues to work on regards this PR?

@adamsaghy, I need a response to this, please. Thanks!

adamsaghy commented:
> > > Additional issues:
> > > 
> > > * Liquibase scripts are missing (to add the new column to the savings account transaction database table)
> > > * Testing is missing
> > > * Adding the new field to the supported parameter list is missing
> > > * Validation whether the provided external id length is valid is missing
> > 
> > 
> > @adamsaghy are these additional issues to work on regards this PR?
> 
> @adamsaghy, I need a response to this, please. Thanks!

I am afraid there are fundamental issues here:
- All irrelevant changes should be removed, like changes of `config/docker/env/aws.env` or `gradlew`
- External id is either provided by customer ("externalId" field value was provided by the incoming request)
- Or we need to check the value of `enable-auto-generated-external-id` global configuration: in case it's true, we need to generate a random value to it -> use `ExternalIdFactory` (it's do the check and generation too!)
- This should be handled in the service and provide the value to the entities
- No hardcoded `null` value, no hardcoded `ExternalId.empty()` (unless external id generation is not enabled and customer was not provided external id either)
- Do not inject services / components into any DTO or JPA entities!
-
